conda_runner.go

```diff
@@ -0,0 +1,171 @@
+type Summer interface {
+	Sum(arg ...string) (string, error)
+}
+
+// CondaRunner implements the Runner interface.
+type CondaRunner struct {
+	logger     scribe.Logger
+	executable Executable
+	summer     Summer
+}
+
+// NewCondaRunner creates an instance of CondaRunner given an Executable, a Summer, and a Logger.
+func NewCondaRunner(executable Executable, summer Summer, logger scribe.Logger) CondaRunner {
+	return CondaRunner{
+		executable: executable,
+		summer:     summer,
+		logger:     logger,
+	}
+}
+
+// ShouldRun determines whether the conda environment setup command needs to be
+// run, given the path to the app directory and the metadata from the
+// preexisting conda-env layer. It returns true if the conda environment setup
+// command must be run during this build, the SHA256 of the package-list.txt in
+// the app directory, and an error. If there is no package-list.txt, the sha
+// returned is an empty string.
+func (c CondaRunner) ShouldRun(workingDir string, metadata map[string]interface{}) (run bool, sha string, err error) {
+	lockfilePath := filepath.Join(workingDir, LockfileName)
+	_, err = os.Stat(lockfilePath)
+
+	if errors.Is(err, os.ErrNotExist) {
+		return true, "", nil
+	}
+
+	if err != nil {
+		return false, "", err
+	}
+
+	updatedLockfileSha, err := c.summer.Sum(lockfilePath)
+	if err != nil {
+		return false, "", err
+	}
+
+	if updatedLockfileSha == metadata[LockfileShaName] {
+		return false, updatedLockfileSha, nil
+	}
+
+	return true, updatedLockfileSha, nil
+}
+
+// Execute runs the conda environment setup command and cleans up unnecessary
+// artifacts. If a vendor directory is present, it uses vendored packages and
+// installs them in offline mode. If a packages-list.txt file is present, it creates a
+// new environment based on the packages list. Otherwise, it updates the
+// existing packages to their latest versions.
+//
+// For more information about the commands used, see:
+// https://docs.conda.io/projects/conda/en/latest/commands/create.html
+// https://docs.conda.io/projects/conda/en/latest/commands/update.html
+// https://docs.conda.io/projects/conda/en/latest/commands/clean.html
+func (c CondaRunner) Execute(condaLayerPath string, condaCachePath string, workingDir string) error {
+	vendorDirExists, err := fileExists(filepath.Join(workingDir, "vendor"))
+	if err != nil {
+		return err
+	}
+
+	lockfileExists, err := fileExists(filepath.Join(workingDir, LockfileName))
+	if err != nil {
+		return err
+	}
+
+	args := []string{
+		"create",
+		"--file", filepath.Join(workingDir, LockfileName),
+		"--prefix", condaLayerPath,
+		"--yes",
+		"--quiet",
+	}
+
+	if vendorDirExists {
+
+		vendorArgs := []string{
+			"--channel", filepath.Join(workingDir, "vendor"),
+			"--override-channels",
+			"--offline",
+		}
+		args = append(args, vendorArgs...)
+
+		c.logger.Subprocess("Running conda %s", strings.Join(args, " "))
+
+		err = c.executable.Execute(pexec.Execution{
+			Args: args,
+		})
+		if err != nil {
```

Comment:
There should output of both the stdout and stderr in the event of a command failure. This will help the user debug the problem they are having.

Reply:
Resolved